### PR TITLE
EP-2420 - Retain GA query params

### PR DESCRIPTION
**Pull request by wrandall22**
[Jira](https://jira.cru.org/browse/EP-2420)

The Google Analytics flow is not properly capturing data and we're trying to figure out why. While testing, I noticed we lose any query parameters after clicking either button in the nav cart. This includes any `utm_???` and `ga_###` type query parameters. This seems like it could be one of the leading causes of the issue.

### CODE REVIEW

File: src/app/signIn/signIn.component.js [hunk 1]
```diff
@@ -39,10 +39,10 @@ class SignInController {
   checkoutAsGuest () {
     this.sessionService.downgradeToGuest(true).subscribe({
       error: () => {
-        this.$window.location = '/checkout.html'
+        this.$window.location = `/checkout.html${window.location.search}`
       },
       complete: () => {
-        this.$window.location = '/checkout.html'
+        this.$window.location = `/checkout.html${window.location.search}`
```
Comment (wrandall22) [inline src/app/signIn/signIn.component.js:45]: Looks like this one doesn't work right.

**Comment by wrandall22** [comment]
Currently being tested by analytics team

File: src/common/components/nav/navCart/navCart.component.spec.js [hunk 2]
```diff
@@ -133,4 +133,16 @@ describe('navCart', () => {
...
     })
   })
+
+  describe('buildCartUrl', () => {
+    it('should build a url without query parameters', () => {
+      expect($ctrl.buildCartUrl()).toEqual('https://give-stage2.cru.org/cart.html')
+    })
+
+    it('should build a url with query parameters', () => {
+      delete window.location
```
Comment (canac) [inline src/common/components/nav/navCart/navCart.component.spec.js:143]: Is this `delete` line needed here and in `src/common/services/api/cart.service.spec.js`?

File: src/app/cart/cart.component.js [hunk 3]
```diff
@@ -103,7 +103,7 @@ class CartController {
   }
 
   checkout () {
-    this.$window.location = this.sessionService.getRole() === 'REGISTERED' ? '/checkout.html' : '/sign-in.html'
+    this.$window.location = this.sessionService.getRole() === 'REGISTERED' ? `/checkout.html${window.location.search}` : `/sign-in.html${window.location.search}`
```
[the remark this answers is not in the document]
Comment (wrandall22) [reply src/app/cart/cart.component.js:106]: The issue with that is that it doesn't handle query strings the same way. With `$window.location` you can't just get the full query string, you can get an array of the query parameters, which you then have to turn into a query string manually.

File: src/common/components/nav/navCart/navCart.component.spec.js (hunk 2 above)
Comment (wrandall22) [reply src/common/components/nav/navCart/navCart.component.spec.js:143]: Yeah, this is what allows us to set `window.location` in the next line.

File: src/app/cart/cart.component.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (wrandall22) [reply src/app/cart/cart.component.js:106]: A benefit with using `$window.location` is that it is a bit nicer to mock for tests.

File: src/common/components/nav/navCart/navCart.component.spec.js (hunk 2 above)
Comment (canac) [reply src/common/components/nav/navCart/navCart.component.spec.js:143]: OK. I would have thought you could just overwrite `window.location` like you're doing on the next line.

File: src/app/cart/cart.component.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (canac) [reply src/app/cart/cart.component.js:106]: I saw in the docs that `$location` converts the query string to an object, but I just console logged `$window.location` and it's a normal `window.location` object.

[the remark this answers is not in the document]
Comment (wrandall22) [reply src/app/cart/cart.component.js:106]: Ok, I'll give it a shot.

**Review by canac — APPROVED** [review]

**Pull request merged**